gc.c

```diff
@@ -5180,16 +5183,27 @@ check_stack_for_moved(rb_objspace_t *objspace)
     each_machine_stack_value(ec, revert_machine_stack_references);
 }
 
-static void
-gc_compact_finish(rb_objspace_t *objspace, rb_size_pool_t *pool, rb_heap_t *heap)
+static bool
+gc_uncompacted_pools_p(rb_objspace_t *objspace)
 {
     for (int i = 0; i < SIZE_POOL_COUNT; i++) {
         rb_size_pool_t *size_pool = &size_pools[i];
-        rb_heap_t *heap = SIZE_POOL_EDEN_HEAP(size_pool);
-        gc_unprotect_pages(objspace, heap);
+        if (!size_pool->compacted) {
+            return TRUE;
+        }
     }
+    return FALSE;
+}
 
-    uninstall_handlers();
+static void
+gc_compact_finish_pool(rb_objspace_t *objspace, rb_size_pool_t *pool, rb_heap_t *heap)
+{
+    pool->compacted = TRUE;
+
+    gc_unprotect_pages(objspace, heap);
```

Comment:
I don't think you can unprotect pages before compaction is completed. Objects in other size pools may refer to objects in this size pool and refer to it before their references get updated.

Reply:
Never mind, it looks you update references after it so it should be fine.